apiserver/facades/client/client/status.go

```diff
@@ -274,6 +281,22 @@ func (c *Client) FullStatus(args params.StatusParams) (params.FullStatus, error)
 				return noStatus, errors.Annotate(err, "could not filter applications")
 			} else if !matches {
 				delete(context.applications, appName)
+				// delete relations for this app
+				if appRels, ok := context.relations[appName]; ok {
+					deleteRelations(appName, appRels)
+				}
```

Comment:
if we are doing these deletions, why do we need to do the later deletions. Shouldn't we have already handled everything?

Reply:
The idea is to optimise later filtering indeed. This particular section deals with matching on units and if we encounter situations where we need to eliminate an application, we'd remove its relations here as well.

However, as with all optimisations - this is just a fall back. The proper place to deal with relations filtering is n the separate section below.

I am happy (albeit reluctant) to remove this code from unit filtering. It does not hurt.

Reply:
This is the only place that will remove relations now - whenever we determine that an application does not fit the filter, we remove its relations from the returned result too.